ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt

```diff
@@ -571,19 +547,142 @@ private sealed class RuleExpr<out T> {
         }
 
         @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
-        private fun wrapRangeInCollection(range: LongRange): Collection<Long> {
-            if (range.first > range.last) {
+        private fun <T : Comparable<T>> wrapRangeInCollection(
+            range: ClosedRange<T>
+        ): Collection<T> {
+            if (range.start > range.endInclusive) {
                 throw IllegalArgumentException("can't use inverted range")
             }
 
-            return object : ClosedRange<Long> by range, Collection<Long> {
-                override val size: Int get() = (endInclusive - start).toInt()
-                override fun containsAll(elements: Collection<Long>) = elements.all {
+            return object : ClosedRange<T> by range, Collection<T> {
+                override fun containsAll(elements: Collection<T>) = elements.all {
                     range.contains(it)
                 }
-                override fun contains(value: Long) = range.contains(value)
-                override fun iterator() = range.iterator()
+                override fun contains(value: T) = range.contains(value)
+                override val size: Int get() {
+                    TODO("Doesn't need this function")
+                }
+                override fun iterator(): Iterator<T> {
+                    TODO("Doesn't need this function")
+                }
+            }
+        }
+    }
+}
+
+internal class Ip {
+    internal data class Numeric(val ipString: String) : Comparable<Numeric> {
+        val value: BigInteger by lazy {
+            val segments = ipString.expandIpv6().toSegments()
+
+            if (segments.size > 4) {
+                segments.fold(BigInteger.fromInt(0)) { acc, seg -> (acc shl 16) + seg.toInt() }
+            } else {
+                segments.fold(BigInteger.fromInt(0)) { acc, octet -> (acc shl 8) + octet.toShort() }
+            }
+        }
+
+        override fun compareTo(other: Numeric) = value.compareTo(other.value)
+    }
+
+    internal data class IpRange(val cidr: String) : ClosedRange<Numeric> {
+        private lateinit var min: Numeric
+        private lateinit var max: Numeric
+
+        init {
+            val ipWidth = cidr.split('/')
+            val segments = ipWidth[0].expandIpv6().toSegments()
+
+            if (segments.size > 4) {
+                val width = ipWidth.getOrNull(1)?.toInt() ?: 128
+                calculateIpv6Range(segments, width)
+            } else {
+                val width = ipWidth.getOrNull(1)?.toByte() ?: 32
+                calculateIpv4Range(segments, width)
+            }
+        }
+
+        private fun calculateIpv6Range(segments: List<UShort>, width: Int) {
```

Comment:
I assume we're using `UShort` to represent IPv6's quartets. Did you consider using `UByte` and representing IPv6 using its octets instead, so that the underlying logic is the same across both versions?

Reply:
This is an interesting point, I haven't thought of representing IPv6 in octets form before this discussion.

I was so fixated on `FFFF:FFFF:FFFF:FFFF:FFFF:FFFF:FFFF:FFFF` for IPv6, and didn't realize that it's just a formatting. It can also be said like `255.255.255.255.255.255.255.255.255.255.255.255.255.255.255.255` which is essentially IPv4 times 4